[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108528

**This PR is a 99% copy paste of Sam Gross** (@colesbury) work at https://github.com/pytorch/pytorch/pull/100642. Copied from there

--------
The NN_MODULE guard now subsumes guards on Module attributes. The check_fn will fail if the module attributes are changed (such as Module.training), parameters, submodules, and buffers are added or removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed the check_fn fails -- for faster overall checks.

-----

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

anijain2305 committed:
[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented on test/dynamo/test_modules.py:
```diff
@@ -1711,7 +1712,111 @@ def fn(x):
+        run_test_case("register_full_backward_hook", backward_hook, 6)
+        run_test_case("register_full_backward_pre_hook", backward_pre_hook, 10)
+
+    def test_unspecialized_nn_module(self):
```
This is the reason of cm3leon graph breaks change.

anijain2305 commented on benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_dynamic_inference.csv:
```diff
@@ -10,7 +10,7 @@ basic_gnn_gcn,pass,6
 basic_gnn_sage,pass,0
 clip,pass_due_to_skip,0
-cm3leon_generate,pass,6
+cm3leon_generate,pass,8
```
This is benign. Please look at this test why this number is increased - https://github.com/pytorch/pytorch/pull/108528/files#r1315372462

anijain2305 commented on torch/_dynamo/guards.py:
```diff
@@ -430,22 +430,36 @@ def CONSTANT_MATCH(self, guard: Guard):
             self.EQUALS_MATCH(guard)
 
     def NN_MODULE(self, guard: Guard):
+        # TODO(janimesh) - This id_match can be removed because nn_module_guard
+        # checks this in C. However, we need this redundant check to allow
+        # flexible cache size policy when we guard on self of nn module
+        # instances. See Note in torch/_dynamo/cache_size.py
         self.ID_MATCH(guard)
```
@colesbury This is a change from your PR. I need this because we want to limit the number of cache entries per ID_MATCH'd object. This ID_MATCH is now hidden inside the C `nn_module_guard` where we do `mod == saved_mod` check.

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"


This PR is a 99% copy paste of Sam Gross (colesbury) work at https://github.com/pytorch/pytorch/pull/100642. Copied from there

--------
The NN_MODULE guard now subsumes guards on Module attributes. The check_fn will fail if the module attributes are changed (such as Module.training), parameters, submodules, and buffers are added or removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed the check_fn fails -- for faster overall checks.

-----

I was unable to push the changes to that PR because its in Sam's personal fork. So, I copied it here. Perhaps we can carry on with the reviews here for a week, and I can move this code to his PR when he comes back from vacation (Sep 11).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"


**This PR is a 99% copy paste of Sam Gross** (colesbury) work at https://github.com/pytorch/pytorch/pull/100642. Copied from there

--------
The NN_MODULE guard now subsumes guards on Module attributes. The check_fn will fail if the module attributes are changed (such as Module.training), parameters, submodules, and buffers are added or removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed the check_fn fails -- for faster overall checks.

-----

I was unable to push the changes to that PR because its in Sam's personal fork. So, I copied it here. Perhaps we can carry on with the reviews here for a week, and I can move this code to his PR when he comes back from vacation (Sep 11).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented on benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_dynamic_training.csv:
```diff
@@ -5,11 +5,10 @@ LearningToPaint,pass,8
 alexnet,pass,8
 attention_is_all_you_need_pytorch,pass,8
-basic_gnn_edgecnn,pass,23
+basic_gnn_edgecnn,pass,25
```
Same as https://github.com/pytorch/pytorch/pull/108528/files#r1315372462

anijain2305 commented on benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_dynamic_training.csv:
```diff
@@ -20,7 +19,7 @@ hf_Albert,pass,7
 hf_DistilBert,pass,7
 hf_GPT2,pass,7
-hf_Reformer,pass,27
+hf_Reformer,pass,30
```
Same as https://github.com/pytorch/pytorch/pull/108528/files#r1315372462

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -495,12 +496,157 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+typedef struct {
+  PyDictObject* dict; // borrowed reference
+  uint64_t dict_version_tag;
+} AttrTag;
```
some docs here would be nice?

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"


**This PR is a 99% copy paste of Sam Gross** (colesbury) work at https://github.com/pytorch/pytorch/pull/100642. Copied from there

--------
The NN_MODULE guard now subsumes guards on Module attributes. The check_fn will fail if the module attributes are changed (such as Module.training), parameters, submodules, and buffers are added or removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed the check_fn fails -- for faster overall checks.

-----

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on test/dynamo/test_modules.py:
```diff
@@ -1711,7 +1712,111 @@ def fn(x):
+        # marked the self to be UnspecializedNNModuleVariable (because of self.a
+        # in the second graph), and therefore during the recompilation, we do
+        # not introduce any nn_module_guard. So, in all we have just one
+        # recompilation.
```
Wouldn't it be better to immediately kill the NNModuleVariable immediately when self.a changes in Dynamo? This would save you an unnecessary compile at the start.

anijain2305 replied:
We kill the NNModuleVariable immediately on self.a mutation. But thats not enough where there is a graph break, and the mutation happens in the second graph. The first graph does not know that there will be mutation, so it assumes NNModuleVariable.

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -523,12 +669,22 @@ PyObject* torch_c_dynamo_guards_init() {
-  if (m == nullptr)
-    return nullptr;
+  if (PyType_Ready(&NNModuleGuardType) < 0)
+    return NULL;
```
nit: nullptr in C++ files, not NULL

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -495,12 +496,157 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+
+    PyObject* key = PyUnicode_FromString(name);
+    if (key == NULL) {
+      return NULL;
```
Need to deallocate guard on failure

ezyang replied:
and dict I guess too

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -430,22 +430,36 @@ def CONSTANT_MATCH(self, guard: Guard):
+            # TorchDynamo graph breaks on LSTMs, but this is a way if user wants
+            # to override it. LSTMs change the module state in every invocation,
+            # leading to recompilations.
+            return
```
You need to warn that RNN guards are ignored in this case right?

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -430,22 +430,36 @@ def CONSTANT_MATCH(self, guard: Guard):
+            # we might be trying to install a guard before a super().__init__() call when
+            # the module is missing _parameters, _modules, and other attributes.
+            # For now, we skip installing the guard.
+            return
```
Also warning here right?

ezyang reviewed (approved):
Thanks for finishing this off.

anijain2305 committed:
Update on "[dynamo][finishing colesbury's PR 100642] Guard on nn.Module dicts and type"


**This PR is a 99% copy paste of Sam Gross** (colesbury) work at https://github.com/pytorch/pytorch/pull/100642. Copied from there

--------
The NN_MODULE guard now subsumes guards on Module attributes. The check_fn will fail if the module attributes are changed (such as Module.training), parameters, submodules, and buffers are added or removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed the check_fn fails -- for faster overall checks.

-----

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it has some nasty merge conflicts after the revert of D48910794.  I need to revert this so the conflict could be resolved.  Please help rebase this tomorrow and reland the change' -c other

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it has some nasty merge conflicts after the revert of D48910794. I need to revert this so the conflict could be resolved. Please help rebase this tomorrow and reland the change' -c weird

huydhn commented:
@pytorchbot rebase

huydhn commented:
@pytorchbot merge

huydhn commented:
@anijain2305 Could you help resolve the conflict and reland this change? This was a nasty merge conflicts, so I had no choice but to reverted this one to unblock diff train

voznesenskym reviewed (changes requested):
Glad to see our work crossing the finish line :) please visit `guard_nn_modules` flag and remove it as needed.

edit: We are still accruing guards all over for where `guard.is_nn_module()` - we probably don't need to do that anymore with this change, right? So we should just not do that, in favor of the new guards, and pull the bool from config and the installation time exclusion it provides.

anijain2305 commented:
Reland PR here - https://github.com/pytorch/pytorch/pull/108883
